Add extra event metadata

dpb587 opened the pull request:
Add additional metadata fields...
- `@ingestor.service` - arbitrary input method (e.g. `lumberjack`, `syslog`)
- `@ingestor.job` - the job name and index from whichever ingestor received the event (e.g. `ingestor/0`)
- `@ingestor.timestamp` - timestamp the ingestor received the event
- `@parser.job` - the job name and index from whichever parser processed the event
- `@parser.timestamp` - timestamp the parser received the event
- `@parser.duration` - time (ms) the parser spent parsing the message
- `@timer.emit_to_ingested` - time delta (ms) between when the event was emitted and received by the ingestor
- `@timer.emit_to_parsed` - time delta (ms) between when the event was emitted and fully parsed
- `@timer.ingested_to_parsed` - time delta (ms) between when the event was ingested and fully parsed (essentially the total time logsearch was involved)

The `@timer.emit_` values make several assumptions - namely that the log emitter has an accurate time. Additionally, logs that don't provide millisecond times will generally hover ~500 ms due to `ms - s` deltas.

dpb587 committed:
Update bosh-lite with new director UUID and quieter CPU footprint

dpb587 committed:
Add numerous additional metadata fields

 * `@ingestor.service` - arbitrary input method (e.g. `lumberjack`, `syslog`)
 * `@ingestor.job` - the job name and index from whichever ingestor received the event (e.g. `ingestor/0`)
 * `@ingestor.timestamp` - timestamp the ingestor received the event
 * `@parser.job` - the job name and index from whichever parser processed the event
 * `@parser.timestamp` - timestamp the parser received the event
 * `@timer.parser` - time (ms) the parser spent parsing the message
 * `@timer.event_lag` - time delta (ms) between when the event was emitted and fully parsed
 * `@timer.ingestion_lag` - time delta (ms) between when the event was emitted and received by the ingestor

dpb587 commented:
This is currently deployed to meta.

dpb587 commented:
I feel like this is going to be adding _so_ much generally-unnecessary data. When initially discussing, I asked whether this could include a config/property flag to enable/disable; I'd like to re-request supporting that flag.

mrdavidlaing commented:
@dpb587, I made [a dashboard to analyse these metrics](http://api.meta.logsearch.io/_plugin/kibana/index.html#/dashboard/elasticsearch/Pipeline%20metrics) on the meta cluster

And here is a sample log event:

``` json
{
  "_index": "logstash-2014.10.16",
  "_type": "elasticsearch_request",
  "_id": "W8VJGjy6RHS-f0RkzzbcUQ",
  "_score": null,
  "_source": {
    "@version": "1",
    "@timestamp": "2014-10-16T07:50:51.092Z",
    "@ingestor.remote_host": "10.12.0.51:45250",
    "@ingestor.service": "syslog",
    "@ingestor.job": "ingestor/0",
    "@ingestor.timestamp": "2014-10-16T07:50:51.422+00:00",
    "@parser.job": "log_parser/2",
    "@parser.timestamp": "2014-10-16T07:50:51.475+00:00",
    "@type": "elasticsearch_request",
    "@message": "{\"time\":\"2014-10-16T07:50:51.092Z\",\"starttime\":\"2014-10-16T07:50:51.092Z\",\"localaddr\":\"127.0.0.1\",\"localport\":9200,\"remoteaddr\":\"127.0.0.1\",\"remoteport\":60266,\"scheme\":\"http\",\"method\":\"GET\",\"path\":\"/logstash-2014.10.16/_aliases\",\"querystr\":\"ignore_missing=true\",\"code\":200,\"status\":\"OK\",\"size\":38,\"duration\":0,\"year\":\"2014\",\"month\":\"10\",\"day\":\"16\",\"hour\":\"07\",\"minute\":\"50\",\"dow\":\"Thu\",\"cluster\":\"live-logsearch\"}",
    "received_at": "2014-10-16 07:50:51 +0000",
    "received_from": "%{host}",
    "tags": [
      "syslog_standard"
    ],
    "@shipper": {
      "host": "946ad785-e60c-4200-9683-82d5db87afa6"
    },
    "@source": {
      "path": "/var/vcap/sys/log/elasticsearch/requests.log",
      "bosh_director": "bosh-meta-logsearch-io",
      "bosh_deployment": "meta-logsearch-io",
      "bosh_job": "api/0",
      "bosh_template": "elasticsearch"
    },
    "time": "2014-10-16T07:50:51.092Z",
    "starttime": "2014-10-16T07:50:51.092Z",
    "localaddr": "127.0.0.1",
    "localport": 9200,
    "remoteaddr": "127.0.0.1",
    "remoteport": 60266,
    "scheme": "http",
    "method": "GET",
    "path": "/logstash-2014.10.16/_aliases",
    "querystr": "ignore_missing=true",
    "code": 200,
    "status": "OK",
    "size": 38,
    "duration": 0,
    "cluster": "live-logsearch",
    "message_data": {},
    "@timer.parser": 3,
    "@timer.event_lag": 386,
    "@timer.ingestion_lag": 56
  }
}
```

mrdavidlaing commented:
Some thoughts:
1.  This is fantastically useful data for measuring ( and diagnosing problems with ) ingestion throughput
2.  Because you can't trust the source event clock (and it will be worse when it comes from Windows); I've noticed [cases where event_lag is negative](http://api.meta.logsearch.io/_plugin/kibana/index.html#dashboard/temp/WJtblHb7TViSi2Cpl7EzkA).  I think the `*_lag` metric will cause more confusion than good; unless we redefine it to use it's starting point as `@ingestor.timestamp`
3.  I think we should group the metadata into "objects" like we do with `@source`, eg:

``` json
@ingestor { 
     "remote_host": "10.12.0.51:45250",
     "service": "syslog",
     "job": "ingestor/0",
     "timestamp": "2014-10-16T07:50:51.422+00:00"
}
```

(4)  Rather than an "on/off" flag, lets go with a debug level flag:
-  `NONE`:  no metadata
-  `INFO`: very basic metadata - perhaps just the `@timer`?
-  `DEBUG`:  everything

To start with, lets just implement `NONE`/`DEBUG`; and then we can "upgrade" that once we know what should go in `INFO` - i.e, the level of metadata that you need to monitor whether there is a problem (but not where the problem is)

dpb587 committed:
Switch to ingestor/parser/timer objects; rename timer events; fix emit_to_ingested calculation

dpb587 committed:
Add logstash.metadata_level property to enable/disable extra fields

dpb587 commented:
I've made the following changes:
- switched the fields to objects
- added a `logstash.metadata_level` property with `NONE`/`DEBUG` values (may want to rename the property)
- moved the `@timer.parser` to `@parser.duration`
- renamed the `@timer` fields to `emit_to_ingested`, `emit_to_parsed`, `ingested_to_parsed` (new)

If we're going to support multiple metadata levels, I'm thinking the following...
- `NONE` - nothing extra
- `INFO` - add the `*.timestamp` fields
- `DEBUG` - add the `@timer.*` fields with those time calculations
- `TRACE` - add the `*.job` fields

I think `*.job` is least useful and, therefore least likely to be used, because all nodes should be running under the same configuration. The `@timer` tasks are simply adding CPU overhead for the purpose of being able to automate and graph delta/processing rates. The `*.timestamp` fields being most useful for manual diagnosis and graphing inbound rates.

dpb587 committed:
Add shared Kibana dashboard for metrics

dpb587 commented:
This failed on production with the following error:

```
{
  :timestamp=>"2014-10-20T18:14:27.184000+0000",
  :message=>"Exception in filterworker",
  "exception"=>#<TypeError: can't dup NilClass>,
  "backtrace"=>[
    "org/jruby/RubyKernel.java:2064:in `dup'",
    "file:/var/vcap/data/packages/logstash/5e1a6fd74024db3c4923397f71da1c71a0cb51d3.1-77343504b4a15b49e874b48842ac11f6c2c8ef4d/logstash/vendor/jar/jruby-complete-1.7.11.jar!/META-INF/jruby.home/lib/ruby/1.9/date/format.rb:833:in `_parse'",
    "file:/var/vcap/data/packages/logstash/5e1a6fd74024db3c4923397f71da1c71a0cb51d3.1-77343504b4a15b49e874b48842ac11f6c2c8ef4d/logstash/vendor/jar/jruby-complete-1.7.11.jar!/META-INF/jruby.home/lib/ruby/1.9/time.rb:265:in `parse'",
    "(ruby filter code):1:in `register'",
    "org/jruby/RubyProc.java:271:in `call'",
    "/var/vcap/packages/logstash/logstash/lib/logstash/filters/ruby.rb:38:in `filter'",
    "(eval):2228:in `initialize'",
    "org/jruby/RubyProc.java:271:in `call'",
    "/var/vcap/packages/logstash/logstash/lib/logstash/pipeline.rb:262:in `filter'",
    "/var/vcap/packages/logstash/logstash/lib/logstash/pipeline.rb:203:in `filterworker'",
    "/var/vcap/packages/logstash/logstash/lib/logstash/pipeline.rb:143:in `start_filters'"
  ],
  :level=>:error
}
```

dpb587 committed:
Use conditionals in parser metadata in case fields are missing

dpb587 committed:
Commit the latest metadata dashboard we have been using

dpb587 commented:
We've deployed this to a couple clusters now and they have been stable. I've committed the latest dashboard which shows the ingestion stats and how performant they are for the various log types (pictured below). It has been immensely useful in helping us track down some of our performance issues and slow log parsers. I think we should go ahead and merge this.

![screen shot 2014-10-24 at 12 11 19 pm](https://cloud.githubusercontent.com/assets/207601/4773295/fc2ae246-5ba0-11e4-811e-70273e14a132.png)
